orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/image/ImageTagger.java

```diff
@@ -96,12 +105,24 @@ protected Collection findImages(Collection<String> imageNames, Stage stage, Clas
   }
 
   @VisibleForTesting
-  Collection<String> upstreamImageIds(Stage sourceStage, String cloudProviderType) {
-    Collection<Stage> imageProvidingAncestorStages = sourceStage.ancestors().stream().filter((Stage stage) -> {
+  Collection<String> upstreamImageIds(Stage sourceStage, Collection<String> consideredStages, String cloudProviderType) {
+    List<Stage> ancestors = sourceStage.ancestors();
+    Collection<Stage> imageProvidingAncestorStages = ancestors.stream().filter((Stage stage) -> {
       String cloudProvider = (String) stage.getContext().getOrDefault("cloudProvider", stage.getContext().get("cloudProviderType"));
```

Comment:
Could you consolidate the logic below into this filter? Something like:
```java
boolean consideredStageRefIdMatches = consideredStages == null || consideredStages.isEmpty() || consideredStages.contains(stage.refId);
(stage.getContext().containsKey("imageId") || stage.getContext().containsKey("amiDetails")) && cloudProviderType.equals(cloudProvider) && consideredStageRefIdMatches
```
Or is there more subtle complexity in the collection of parent refIds below that makes it easier to keep as separate methods?

Reply:
Yeah, this was overly complicated. I've simplified it a lot using your suggestion. Thanks.